FIDO passkey post (#1438)
oktadeveloper/okta-blog: _source/_posts/2023-08-30-passwordless-authentication.md

---
layout: blog_post
title: "We're Living in a Passwordless World"
author: alisa-duncan
by: advocate
communities: [security,mobile,javascript]
description: "From passwords to passwordless — learn about how to elevate your authentication security."
tags: ["passkey", "authentication", "fido"]
tweets:
- ""
- ""
- ""
image: blog/passwordless-authentication/social.jpg
type: awareness
---

Passwords? Bah! Who wants to remember all those letter, number, and special character combinations? Password managers and password manager browser extensions can help us, but what if we stopped using passwords entirely? What if we stopped using authentication factors, such as SMS one-time passwords (OTP), that pose risks where malicious actors can access sensitive authentication information (phishing)? What if authentication was so seamless that we could use built-in device hardware capabilities instead?

> Seamless, easy-to-use authentication is the future many dream of, including a group called the [Fast Identity Online (FIDO) Alliance](https://fidoalliance.org/). The FIDO Alliance is a group of organizations that focuses on providing open and free authentication standards utilizing more secure technologies. And we've heard of the [World Wide Web Consortium (W3C)](https://www.w3.org/) as a group that builds web standards. How do these two groups intersect when it comes to authentication security? Let's better understand the collaboration and what it means for us.

Reviewer: ```suggestion
Seamless, easy-to-use authentication is the future many dream of, including a group called the [Fast Identity Online (FIDO) Alliance](https://fidoalliance.org/). The FIDO Alliance is a group of organizations that focuses on providing open and free authentication standards utilizing more secure technologies. We've also heard the [World Wide Web Consortium (W3C)](https://www.w3.org/) is a group that builds web standards. How do these two groups intersect when it comes to authentication security? Let's better understand the collaboration and what it means for us.
```

{% include toc.md %}

## Authentication factors

Passwords were the default authentication mechanism for years, but passwords are notoriously insecure and [cause most data breaches](https://www.okta.com/resources/whitepaper/protecting-against-data-breaches/). Research shows that >80% of breaches stem from stolen or weak passwords. So it's time for something more secure.

> Passwords are one way to authenticate, but there are other options. Authentication factors break down into the following categories:

Reviewer: ```suggestion
Passwords are just one way to authenticate, but there are other options. Authentication factors break down into the following categories:
```

{% img blog/passwordless-authentication/webauthn-caniuse.jpg alt:"WebAuthn support across browsers showing Internet Explorer and Opera Mini having no support, Firefox desktop and mobile has partial support, and remaining browsers including Chrome, Edge, Safari on desktop and mobile having support." width:"800" %}{: .center-image }

FIDO's Client to Authenticator Protocol complements WebAuthn to create a seamless, easy-to-use, secure authentication experience. 

{% img blog/passwordless-authentication/webauthn-ctap.jpg alt:"Icons representing WebAuthn in the browser and CTAP devices with a heart in between" width:"800" %}{: .center-image }

> A new term for this authentication mechanism without passwords is often called "passwordless." And the combination of WebAuthn and CTAP2 hardware devices to enable passwordless authentication is called a "passkey." What a great name!

Reviewer: I'd expand more on passkey here.  This makes it seem like an accidental evolution of passwordless authentication, whereas "passkey" is a joint effort by Apple, Google, and Microsoft to standardize on the concept of implementing seamless authentication and credential management for apps and websites using WebAuthN.

It is a big enough deal to call out as its own separate section, because FIDO -> WebAuthn -> Passkey is a clear progression of standards.  It also can clarify the immediacy and the reason why this discussion is all the more relevant now.

FIDO and WebAuthn have been around for quite some time now, without much traction in the industry, since there's been little to no motivation for companies to add support.  With Apple putting their foot on the pedal, they're forcing the industry to think about this. Google and Microsoft are having to follow suit, and soon most sites and providers out there will fall behind if they don't follow the trend.  Just as 2FA and MFA weren't very widely adopted until key players began pushing for it (and more to the point, enough data breaches occurred to make it financially viable), Passkey is one of those things that will soon become table-stakes for people expecting a certain experience, particularly in the CIAM markets.

Anyway, I digress.  Either way, I think it's important to call out Passkey itself as a 3rd iteration of the standard, to showcase the importance of adoption, and why it's different from standard WebAuthn.

Reviewer: Oh one more thing, I'd suggest you show those screenshots from within Safari, and not from Chrome. Google has very minimal support for Passkeys right now, while Apple has a much richer experience for it.  If you're writing this from a CIAM perspective, Safari is a clearer win because your private keys are shared across iCloud devices, so a user can seamlessly sign in from their iPhone, Mac, or iPad, either on a native app or on its companion website.

Author: Great feedback and explanation of why passkeys are relevant now, @mikenachbaur-okta!  I will work on this - thanks!

## FIDO2 authentication with WebAuthn and CTAP
 
So how does this work? At a high level, your FIDO2 authentication flow utilizes each of the three factors:
1. Something you know, a PIN to unlock the device
2. Something you have, the authenticator device, which may be a security key or a capable phone
3. Something you are, biometric data such as fingerprint or facial recognition

Before we authenticate, we need an account with compliant credentials, so let's start with the registration process. We'll use a public demo site demonstrating how passkeys work as we step through what's going on in the registration and authentication process. Follow along at Auth0's [webauthn.me](https://webauthn.me/).

There are a couple of components at play. From the user's perspective, they interact with a website and a device like their phone. From a technology perspective, there's WebAuthn and CTAP2. I'll show screenshots of each step, discuss the interaction between the web app and server of the relying party, and describe the hand-off between WebAuthn and CTAP2.

> ## Registering a user for passkeys

Reviewer: I think this section could benefit from some sort of flow diagram. The text makes a great deal of sense, but most people will just skim it in my experience.

If anything, I think these paragraphs could be numbered, each of which referencing a numbered path with a flow diagram demonstrating the interactions between the browser, the authenticator, the authorization server, and the web app on the other end.

The screenshots are certainly useful to demonstrate the user's experience, but if anything I would move the user's experience to its own section. The magic is in the server interactions.  The user's experience is intended to be so streamlined as to be almost irrelevant to the user.

Reviewer: Oh, and from the user's experience, the registration process is the least important. Most important to highlight is the seamless workflow the user has when they sign in again.  This is one of the reasons why I suggest highlighting Safari in these examples, because it's a very beautiful experience.

So perhaps there should be 2 sections in this area: 1) Client-to-server Workflow, and 2) User Experience.

---
layout: blog_post
title: "We're Living in a Passwordless World"
author: alisa-duncan
by: advocate
communities: [security,mobile,javascript]
description: "From passwords to passwordless — learn about how to elevate your authentication security."
tags: ["passkey", "authentication", "fido"]
tweets:
- ""
- ""
- ""
image: blog/passwordless-authentication/social.jpg
type: awareness
---

> Passwords? Bah! Who wants to remember all those letters, numbers, and special character combinations? Password managers and password manager browser extensions can help us, but what if we stopped using passwords instead? What if we stopped using other phishable authentication factors, such as SMS one-time passwords (OTP)? What if authentication was so seamless that we could use built-in device hardware capabilities instead?

Reviewer 2: would it make sense to link "phishable" to a definition of phishing in this sentence, just to catch readers who aren't confident in their understanding of it?

Author: Good call, I'll add this.

> Seamless, easy-to-use authentication is the future many dream of, and a group called the [Fast Identity Online (FIDO) Alliance](https://fidoalliance.org/). The FIDO Alliance is a group of organizations that focuses on providing open and free authentication standards utilizing more secure technologies. And we've heard of the [World Wide Web Consortium (W3C)](https://www.w3.org/) as a group that builds web standards. How do these two groups intersect when it comes to authentication security? Let's better understand the collaboration and what it means for us.

Reviewer 2: read first sentence out loud -- missing verb for fido alliance? or should the "and" be "including"?

{% include toc.md %}

## Authentication factors

Passwords were the default authentication mechanism for years, but passwords are notoriously insecure and [cause most data breaches](https://www.okta.com/resources/whitepaper/protecting-against-data-breaches/). Research shows that >80% of breaches stem from stolen or weak passwords. So it's time for something more secure.

Passwords are one way to authenticate, but there are other options. Authentication factors break down into the following categories:

> <table>

Reviewer 2: 😍 yay tables!

Author: i love images!

<tr>
    <td>{% img blog/passwordless-authentication/factor-knowledge.jpg alt:"Icon of someone thinking" width:"100" %}</td>
    <td markdown="span">
      **Something you know** <br/>
      Knowledge factors include passwords and PINs
    </td>
</tr>
<tr>
    <td>{% img blog/passwordless-authentication/factor-possession.jpg alt:"Icon showing devices such as computer, tablet, and phone" width:"100" %}</td>
    <td markdown="span">
    **Something you have**<br/>
    Possession factors include devices such as smart cards, security keys, phones, and tablets
    </td>
</tr>
<tr>
    <td>{% img blog/passwordless-authentication/factor-inherence.jpg alt:"Icon of a hand with fingerprint" width:"100" %}</td>
    <td markdown="span">
    **Something you are**<br/>
    Inference factors include biometrics such as fingerprints and facial recognition
    </td>
</tr>
</table>

> Because of the security risks associated with passwords, the security community recommended adding additional authentication factors to mitigate the concerns about using passwords alone. Unfortunately, there were still security risks and barriers associated with past recommendations. So that we understand how we get to yet a new recommended method for authentication, let's recap a few of the previous advice and why they weren't the perfect solutions:

Reviewer 2: should 'recommended' / 'recommendations' be links? yes if you've got something in mind, otherwise maybe not

Author: I don't have specific links for each of these on hand, but maybe I can find them? Lower priority though

<table>
<tr>
    <td>{% img blog/passwordless-authentication/warning.jpg alt:"Warning icon" width:"100" %}</td>
    <td markdown="span">
      **Any two-factor authentication (2FA) and multi-factor authentication (MFA)** <br/>
      Both 2FA and MFA can still be insecure. For example, legacy knowledge-based authentication factors such as security questions are weak, insecure, and prone to social engineering. SMS OTP, a possession-based factor, is vulnerable to SIM swapping hacks and is phishable.
    </td>
</tr>
<tr>
    <td>{% img blog/passwordless-authentication/warning.jpg alt:"Warning icon" width:"100" %}</td>
    <td markdown="span">
    **Time-based one-time passwords (TOTP)**<br/>

>     TOTP are the codes you get through software authenticators, such as Okta Verify or Google Authenticator, valid for a short time. While TOTP improves SMS OTP, it can be vulnerable to meddler-in-the-middle (MITM) attacks, displays the code in plaintext and expects the user to transcribe the value accurately and quickly to authenticate.

Reviewer 2: ```suggestion
    TOTP are the codes you get through software authenticators, such as Okta Verify or Google Authenticator, valid for a short time. While TOTP improves SMS OTP, any SMS factor is still vulnerable to meddler-in-the-middle (MITM) attacks, displays the code in plaintext and expects the user to transcribe the value accurately and quickly to authenticate.
```

    </td>
</tr>
<tr>
    <td>{% img blog/passwordless-authentication/confused.jpg alt:"Icon of a person confused" width:"100" %}</td>
    <td markdown="span">
    **Smart cards conforming to FIDO's Universal 2nd Factor (U2F) standard, such as security keys**<br/>

>     Smart cards are a very secure and less human error-prone option because they use a locally stored cryptographically secure key and utilize Bluetooth, USB, or Near Field Communication (NFC) instead of the internet or phone line to transmit the authentication code on behalf of the user. While U2F may be an option for work accounts, users found it cumbersome for personal use.

Reviewer 2: "have" factors are also vulnerable to theft -- if someone has my badge, the door thinks that they are me

Author: True. I was only thinking of authentication security from a software sense, but there is an overreliance in some orgs on physical security only, such as key cards/badges. Will consider adding

    </td>
</tr>
</table>

What we want is a robust, phishing-resistant, easy-to-use authentication mechanism.

## Phishing-resistant authentication

> Phishing-resistant authentication factors include local biometric factors and secure possession-based factors. Using cryptographically unique credentials for each website and new browser capabilities that make secure authentication easy to use is the winning combination!  

Reviewer 2: read final sentence of paragraph aloud -- "using ... easy to use" might need to rephrase

> Using FIDO and W3C's combined forces brings us to a new authentication standard called FIDO2. FIDO2 retained the cryptographically secure authentication key from the previous U2F method. They introduced a new protocol to an existing standard for devices to interoperate with web browsers called [Client to Authenticator Protocol (CTAP2)](https://fidoalliance.org/specs/fido-v2.0-ps-20190130/fido-client-to-authenticator-protocol-v2.0-ps-20190130.html). Devices supporting CTAP2 include those with built-in hardware authenticators and external devices. An example of a device with built-in hardware authenticators includes laptops with fingerprint sensors or cameras for facial recognition. External devices include security keys, phones, and watches with Bluetooth, USB, or NFC communication capability.

Reviewer 2: "Devices supporting CTAP2 include those with built-in hardware authenticators and external devices." i am confused, built-in external devices? those with external devices?

Author: yeah, that might be an aggressive grammarly agreement that became unhelpful. I mean devices with built-in biometric scanning hardware and external devices like smart cards or smart keys.

I can find a way to clarify that

{% img blog/passwordless-authentication/ctap.jpg alt:"Graphic of device icons such as phone, laptop, smart card, and smart key along with forms of communication (Bluetooth, USB, and NFC)" width:"800" %}{: .center-image }

> W3C introduced a new web standard for web authentication called [WebAuthn](https://www.w3.org/TR/webauthn-2/). A new browser API, the Web Authentication API, extends from the Credential Management API. The Web Authentication API supports secure authentication credentials using asymmetric encryption (where you have a public and private key) instead of passwords. All modern browsers have full or partial WebAuthn capabilities. You can check which browsers and versions support WebAuthn on [caniuse.com](https://caniuse.com/?search=webauthn). The screen capture shows browser support at the time of publishing. 

Reviewer 2: i hate how the N of WebAuthn is not capitalized. So hard to read. Not your fault, but it's annoying.

Reviewer: Huh I've always capitalized the "N" when I've written it. It seemed so obvious, I never bothered to look up the "official" capitalization. 😠

Reviewer: I'd also like to suggest small tweaks to the first part of the first sentence.  It doesn't illustrate the timeline between FIDO, and the W3C introduction of WebAuthn.

Something like: "The W3C then introduced a new web standard for authentication on the web, called WebAuthn"...

Reviewer: ```suggestion
The W3C then introduced a new web standard for web authentication on the web, called [WebAuthn](https://www.w3.org/TR/webauthn-2/). It added a new browser-based API, the Web Authentication API, which extends from the existing Credential Management API. Instead of using passwords, the Web Authentication API uses secure authentication credentials using asymmetric encryption (where you have a public and private key). All modern browsers now have full or partial support for WebAuthn. You can check which browsers and versions support WebAuthn on [caniuse.com](https://caniuse.com/?search=webauthn). The screen capture shows browser support at the time of publishing. 
```

## Registering a user for passkeys

The user opens their favorite site in the browser. In the FIDO2 world, the site that makes the WebAuthn calls is the _Relying Party_, and the browser is the _Client_. If your site uses a single-page application frontend, the relying party includes both the frontend and backend server. 

{% img blog/passwordless-authentication/webauthnio.jpg alt:"webauthn.io site showing a user name field an buttons to register or authenticate" width:"800" %}{: .center-image }

The web app for the site displays a form to gather user information, an email address, or a user name. The user enters the required information and presses **Register**. The web app recognizes the browser as FIDO2 capable and asks if the user wants to sign in using a passkey instead of a password. Of course, the user says, "Yes!"

{% img blog/passwordless-authentication/webauthnio-register.jpg alt:"webauthn.io site showing a user name entered in the form and a modal from the browser asking to create a passkey" width:"800" %}{: .center-image }

> The web app calls the server with credential request options, such as the username and registration-related data. You can open debugging capabilities in your browser and watch the network requests. Here, we see the payload for the call, which includes the username.

Reviewer 2: can open debugging capabilities be a link to how to open the developer console somewhere? if we catch someone in the lucky 10k of devtools they may have a rough time figuring out what "open debugging capabilities" means

Author: It depends on the browser, so hence, no explicit instructions or links

{% img blog/passwordless-authentication/webauthnio-register-payload.jpg alt:"webauthn.io payload to register a user includes the username along with registration metadata configurable in advanced settings, such as encryption algorithm" width:"800" %}{: .center-image }

The server responds with information about the relying party, a user ID, and a cryptographically random value for the registration process called a "challenge." 

{% img blog/passwordless-authentication/webauthnio-register-response.jpg alt:"webauthn.io response includes relying party id, user name and id, and the challenge" width:"800" %}{: .center-image }

The user selects a device for the passkey, like a phone or tablet. The web app uses WebAuthn to generate a new credential, sending the response information from the server, including the challenge. The credential call from the browser to the authenticator is possible via CTAP2.

> The authenticator takes over, prompting the user for consent and biometric information. The authenticator generates a public/private key set for the website. The authenticator signs the public key with an attestation certificate for the device hardware. Signing the key with the attestation certificate proves the keyset is from a specific device. The authenticator returns the public key in the WebAuthn credential response.

Reviewer 2: are keypairs referred to as key sets in WebAuthn? I haven't heard pairs called sets before but I may be missing context

Reviewer: Good point, "key sets" are often used within OAuth2 to refer to an Authorization Server's list of available public keys (e.g. the `/oauth2/v1/keys` endpoint).  I'd use "key pair" here.

{% img blog/passwordless-authentication/webauthnio-ctap.jpg alt:"webauthn.io site with a modal to create a passkey showing a QR code" width:"800" %}{: .center-image }

The web app calls the server with the public key and the user ID to store the user's credentials.

{% img blog/passwordless-authentication/webauthnio-register-success-payload.jpg alt:"webauthn.io site network call payload to complete registration" width:"800" %}{: .center-image }

> The server verifies the key for completeness and integrity, such as ensuring the certificate's validity, the website origin is accurate, and that the challenge value matches the original response. Once verified, the server persists this info, matching your username with the public key so it has the information needed to authenticate you for the passkey later.

Reviewer 2: "ensuring certificate's validity" is probably the right level of abstraction for here but it makes me curious which actual checks are involved -- check expiry? look up in a CA? look up in system CA, or a specific one?

## Authenticating a user with passkeys

We can see what user authentication looks like with an account for the site set up and the initial passkey created.

The user opens their favorite website (the Relying Party) in their browser (the Client) and initiates signing in by entering their username and pressing **Authenticate**. They'll see a prompt from the client to complete authentication using the passkey they created during registration.

{% img blog/passwordless-authentication/webauthnio-authn.jpg alt:"webauthn.io site authenticate user prompt from the browser" width:"800" %}{: .center-image }

But we want to see what's going on under the covers, so let's look at the network call. The web app calls the server with the username from the user entry. 

{% img blog/passwordless-authentication/webauthnio-authn-request.jpg alt:"webauthn.io site network call request to authenticate user" width:"800" %}{: .center-image }

> The server remembers you're using passkeys! 🎉 The server responds with a challenge and additional information required for authentication, such as Relying Party and user credentials or authenticator information.

Reviewer 2: 🎉

{% img blog/passwordless-authentication/webauthnio-authn-response.jpg alt:"webauthn.io site network call response to authenticate user" width:"800" %}{: .center-image }

> The website utilizes WebAuthn, which causes the browser to pass the challenge to the authenticator. The authenticator verifies it has credentials for the site and can sign the challenge using the keyset registered, a process called "assertion." The authenticator requests user consent via biometric verification or a PIN and returns the signed assertion. The browser presents the user with an authentication challenge through the browser, which we see in the screenshots above of the authentication flow. The user grants consent using the passkey.

Reviewer 2: is there a clearer verb than "utilizes" here?
